File: clients/rust/examples/fake_session.rs
use serde_json::json;
use std::env;
use zmr_client::Client;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut zmr = String::from("zig-out/bin/zmr");
    let mut adb = String::from("tests/fake-adb.sh");
    let mut device = String::from("fake-android-1");
    let mut app_id = String::from("com.example.mobiletest");
    let mut trace_dir = String::from("traces/demo-rust-client");
    let mut trace_out = String::from("traces/demo-rust-client-redacted.zmrtrace");

    let mut args = env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--zmr" => zmr = args.next().ok_or("--zmr requires a value")?,
            "--adb" => adb = args.next().ok_or("--adb requires a value")?,
            "--device" => device = args.next().ok_or("--device requires a value")?,
            "--app-id" => app_id = args.next().ok_or("--app-id requires a value")?,
            "--trace-dir" => trace_dir = args.next().ok_or("--trace-dir requires a value")?,
            "--trace-out" => trace_out = args.next().ok_or("--trace-out requires a value")?,
            _ => return Err(format!("unknown argument: {arg}").into()),
        }
    }

    let mut client = Client::start(
        &zmr,
        [
            "serve",
            "--transport",
            "stdio",
            "--device",
            &device,
            "--app-id",
            &app_id,
            "--adb",
            &adb,
            "--trace-dir",
            &trace_dir,
        ],
    )?;
    let capabilities = client.capabilities()?;
    client.create_session()?;
    client.open_link("exampleapp://rust-client")?;
    client.wait_until(json!({ "text": "Dashboard" }), Some(1000))?;
    client.tap(json!({ "text": "Sign in" }))?;
    client.type_text(
        "agent@example.com",
        Some(json!({ "resourceId": "email-login-email-input" })),
    )?;
    client.assert_not_visible(json!({ "text": "Application has crashed" }), Some(100))?;
    client.assert_healthy(Some(100))?;
    let snapshot = client.snapshot()?;
    let exported = client.export_trace(&trace_out, true, true)?;
    let events = client.trace_events(0, Some(10))?;

    println!(
        "{}",
        json!({
            "protocolVersion": capabilities.protocol_version,
            "activePackage": snapshot.active_package,
            "nodes": snapshot.nodes.len(),
            "events": events.next_seq,
            "traceDir": trace_dir,
            "traceOut": exported.out,
        })
    );

    Ok(())
}

File: clients/kotlin/src/main/kotlin/dev/zmr/FakeSession.kt
package dev.zmr

private data class Options(
    var zmr: String = "zig-out/bin/zmr",
    var adb: String = "tests/fake-adb.sh",
    var device: String = "fake-android-1",
    var appId: String = "com.example.mobiletest",
    var traceDir: String = "traces/demo-kotlin-client",
    var traceOut: String = "traces/demo-kotlin-client-redacted.zmrtrace"
)

fun main(args: Array<String>) {
    val options = parseOptions(args)
    ZmrClient(
        listOf(
            options.zmr,
            "serve",
            "--transport", "stdio",
            "--device", options.device,
            "--app-id", options.appId,
            "--adb", options.adb,
            "--trace-dir", options.traceDir
        )
    ).use { client ->
        val capabilities = client.call("runner.capabilities")
        client.createSession()
        client.call("app.openLink", """{"url":"exampleapp://kotlin-client"}""")
        client.call("wait.until", """{"visible":{"text":"Dashboard"},"timeoutMs":1000}""")
        client.call("ui.tap", """{"selector":{"text":"Sign in"}}""")
        client.call("ui.type", """{"text":"agent@example.com","selector":{"resourceId":"email-login-email-input"}}""")
        client.call("assert.notVisible", """{"selector":{"text":"Application has crashed"},"timeoutMs":100}""")
        client.assertHealthy(timeoutMs = 100)
        val snapshot = client.snapshot()
        val exported = client.call(
            "trace.export",
            """{"out":"${escapeJson(options.traceOut)}","redact":true,"includeScreenshots":true}"""
        )
        val events = client.call("trace.events", """{"afterSeq":0,"limit":10}""")

        println(
            "{" +
                "\"protocolVersion\":\"${extractString(capabilities, "protocolVersion")}\"," +
                "\"activePackage\":\"${extractString(snapshot, "activePackage")}\"," +
                "\"nodes\":${countOccurrences(snapshot, "\"stableId\"")}," +
                "\"events\":${extractNumber(events, "nextSeq")}," +
                "\"traceDir\":\"${escapeJson(options.traceDir)}\"," +
                "\"traceOut\":\"${escapeJson(extractString(exported, "out"))}\"" +
                "}"
        )
    }
}

private fun parseOptions(args: Array<String>): Options {
    val options = Options()
    var index = 0
    while (index < args.size) {
        val value = args.getOrNull(index + 1) ?: error("${args[index]} requires a value")
        when (args[index]) {
            "--zmr" -> options.zmr = value
            "--adb" -> options.adb = value
            "--device" -> options.device = value
            "--app-id" -> options.appId = value
            "--trace-dir" -> options.traceDir = value
            "--trace-out" -> options.traceOut = value
            else -> error("unknown argument: ${args[index]}")
        }
        index += 2
    }
    return options
}

private fun extractString(json: String, key: String): String {
    val match = Regex(""""$key"\s*:\s*"([^"]*)"""").find(json)
    return match?.groupValues?.get(1) ?: ""
}

private fun extractNumber(json: String, key: String): String {
    val match = Regex(""""$key"\s*:\s*([0-9]+)""").find(json)
    return match?.groupValues?.get(1) ?: "0"
}

private fun countOccurrences(value: String, needle: String): Int =
    Regex.escape(needle).toRegex().findAll(value).count()

private fun escapeJson(value: String): String =
    value.replace("\\", "\\\\").replace("\"", "\\\"")
